<a href="https://colab.research.google.com/github/Alex-Witt/DS-Unit-1-Sprint-1-Dealing-With-Data/blob/master/module2-loadingdata/LS_DS_112_Loading_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science - Loading Data

Data comes in many shapes and sizes - we'll start by loading tabular data, usually in csv format.

Data set sources:

- https://archive.ics.uci.edu/ml/datasets.html
- https://github.com/awesomedata/awesome-public-datasets
- https://registry.opendata.aws/ (beyond scope for now, but good to be aware of)

Let's start with an example - [data about flags](https://archive.ics.uci.edu/ml/datasets/Flags).

## Lecture example - flag data

In [0]:
# Step 1 - find the actual file to download

# From navigating the page, clicking "Data Folder"
flag_data_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/flags/flag.data'

# You can "shell out" in a notebook for more powerful tools
# https://jakevdp.github.io/PythonDataScienceHandbook/01.05-ipython-and-shell-commands.html

# Funny extension, but on inspection looks like a csv
!curl https://archive.ics.uci.edu/ml/machine-learning-databases/flags/flag.data

# Extensions are just a norm! You have to inspect to be sure what something is

Afghanistan,5,1,648,16,10,2,0,3,5,1,1,0,1,1,1,0,green,0,0,0,0,1,0,0,1,0,0,black,green
Albania,3,1,29,3,6,6,0,0,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,0,1,0,red,red
Algeria,4,1,2388,20,8,2,2,0,3,1,1,0,0,1,0,0,green,0,0,0,0,1,1,0,0,0,0,green,white
American-Samoa,6,3,0,0,1,1,0,0,5,1,0,1,1,1,0,1,blue,0,0,0,0,0,0,1,1,1,0,blue,red
Andorra,3,1,0,0,6,0,3,0,3,1,0,1,1,0,0,0,gold,0,0,0,0,0,0,0,0,0,0,blue,red
Angola,4,2,1247,7,10,5,0,2,3,1,0,0,1,0,1,0,red,0,0,0,0,1,0,0,1,0,0,red,black
Anguilla,1,4,0,0,1,1,0,1,3,0,0,1,0,1,0,1,white,0,0,0,0,0,0,0,0,1,0,white,blue
Antigua-Barbuda,1,4,0,0,1,1,0,1,5,1,0,1,1,1,1,0,red,0,0,0,0,1,0,1,0,0,0,black,red
Argentina,2,3,2777,28,2,0,0,3,2,0,0,1,0,1,0,0,blue,0,0,0,0,0,0,0,0,0,0,blue,blue
Argentine,2,3,2777,28,2,0,0,3,3,0,0,1,1,1,0,0,blue,0,0,0,0,1,0,0,0,0,0,blue,blue
Australia,6,2,7690,15,1,1,0,0,3,1,0,1,0,1,0,0,blue,0,1,1,1,6,0,0,0,0,0,white,blue
Austria,3,1,84,8,4,0,0,3,2,1,0,0,0,1,0,0,red,0,0,0,0,0,0,0,0,0,0,red,red
Bahamas,1,4,19,0,1,1,0,3,3,0,0,1,1,0,1,0,blue,0,0,

In [0]:
# Step 2 - load the data

# How to deal with a csv? 🐼
import pandas as pd
flag_data = pd.read_csv(flag_data_url)

In [0]:
# Step 3 - verify we've got *something*
flag_data.head()

Afghanistan  5  1   648  16  10  2  0  3  5.1   ...     0.5  0.6  1.6  \
0         Albania  3  1    29   3   6  6  0  0    3   ...       0    0    1   
1         Algeria  4  1  2388  20   8  2  2  0    3   ...       0    0    1   
2  American-Samoa  6  3     0   0   1  1  0  0    5   ...       0    0    0   
3         Andorra  3  1     0   0   6  0  3  0    3   ...       0    0    0   
4          Angola  4  2  1247   7  10  5  0  2    3   ...       0    0    1   

   0.7  0.8  1.7  0.9 0.10  black  green.1  
0    0    0    0    1    0    red      red  
1    1    0    0    0    0  green    white  
2    0    1    1    1    0   blue      red  
3    0    0    0    0    0   blue      red  
4    0    0    1    0    0    red    black  

[5 rows x 30 columns]

In [0]:
# Step 4 - Looks a bit odd - verify that it is what we want
flag_data.count()

Afghanistan    193
5              193
1              193
648            193
16             193
10             193
2              193
0              193
3              193
5.1            193
1.1            193
1.2            193
0.1            193
1.3            193
1.4            193
1.5            193
0.2            193
green          193
0.3            193
0.4            193
0.5            193
0.6            193
1.6            193
0.7            193
0.8            193
1.7            193
0.9            193
0.10           193
black          193
green.1        193
dtype: int64

In [0]:
!curl https://archive.ics.uci.edu/ml/machine-learning-databases/flags/flag.data | wc

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 15240  100 15240    0     0  50969      0 --:--:-- --:--:-- --:--:-- 51140
    194     194   15240


In [0]:
# So we have 193 observations with funny names, file has 194 rows
# Looks like the file has no header row, but read_csv assumes it does
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers:

read_csv(filepath_or_buffer, sep=',', delimiter=None, header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, iterator=False, chunksize=None, compression='infer', thousands=None, decimal=b'.', lineterminator=None, quotechar='"', quoting=0, escapechar=None, comment=None, encoding=None, dialect=None, tupleize_cols=None, error_bad_lines=True, warn_bad_lines=True, skipfooter=0, skip_footer=0, doublequote=True, delim_whitespace=False, as_recarray=None, compact_ints=None, use_unsigned=None, low_memory=True, buffer_lines=None, memory_map=False, float_precision=None)


In [0]:
# Alright, we can pass header=None to fix this
flag_data = pd.read_csv(flag_data_url, header=None)
flag_data.head()

0   1   2     3   4   5   6   7   8   9   ...    20  21  22  \
0     Afghanistan   5   1   648  16  10   2   0   3   5  ...     0   0   1   
1         Albania   3   1    29   3   6   6   0   0   3  ...     0   0   1   
2         Algeria   4   1  2388  20   8   2   2   0   3  ...     0   0   1   
3  American-Samoa   6   3     0   0   1   1   0   0   5  ...     0   0   0   
4         Andorra   3   1     0   0   6   0   3   0   3  ...     0   0   0   

   23  24  25  26 27     28     29  
0   0   0   1   0  0  black  green  
1   0   0   0   1  0    red    red  
2   1   0   0   0  0  green  white  
3   0   1   1   1  0   blue    red  
4   0   0   0   0  0   blue    red  

[5 rows x 30 columns]

In [0]:
flag_data.count()

0     194
1     194
2     194
3     194
4     194
5     194
6     194
7     194
8     194
9     194
10    194
11    194
12    194
13    194
14    194
15    194
16    194
17    194
18    194
19    194
20    194
21    194
22    194
23    194
24    194
25    194
26    194
27    194
28    194
29    194
dtype: int64

In [0]:
flag_data.isna().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
20    0
21    0
22    0
23    0
24    0
25    0
26    0
27    0
28    0
29    0
dtype: int64

### Yes, but what does it *mean*?

This data is fairly nice - it was "donated" and is already "clean" (no missing values). But there are no variable names - so we have to look at the codebook (also from the site).

```
1. name: Name of the country concerned
2. landmass: 1=N.America, 2=S.America, 3=Europe, 4=Africa, 4=Asia, 6=Oceania
3. zone: Geographic quadrant, based on Greenwich and the Equator; 1=NE, 2=SE, 3=SW, 4=NW
4. area: in thousands of square km
5. population: in round millions
6. language: 1=English, 2=Spanish, 3=French, 4=German, 5=Slavic, 6=Other Indo-European, 7=Chinese, 8=Arabic, 9=Japanese/Turkish/Finnish/Magyar, 10=Others
7. religion: 0=Catholic, 1=Other Christian, 2=Muslim, 3=Buddhist, 4=Hindu, 5=Ethnic, 6=Marxist, 7=Others
8. bars: Number of vertical bars in the flag
9. stripes: Number of horizontal stripes in the flag
10. colours: Number of different colours in the flag
11. red: 0 if red absent, 1 if red present in the flag
12. green: same for green
13. blue: same for blue
14. gold: same for gold (also yellow)
15. white: same for white
16. black: same for black
17. orange: same for orange (also brown)
18. mainhue: predominant colour in the flag (tie-breaks decided by taking the topmost hue, if that fails then the most central hue, and if that fails the leftmost hue)
19. circles: Number of circles in the flag
20. crosses: Number of (upright) crosses
21. saltires: Number of diagonal crosses
22. quarters: Number of quartered sections
23. sunstars: Number of sun or star symbols
24. crescent: 1 if a crescent moon symbol present, else 0
25. triangle: 1 if any triangles present, 0 otherwise
26. icon: 1 if an inanimate image present (e.g., a boat), otherwise 0
27. animate: 1 if an animate image (e.g., an eagle, a tree, a human hand) present, 0 otherwise
28. text: 1 if any letters or writing on the flag (e.g., a motto or slogan), 0 otherwise
29. topleft: colour in the top-left corner (moving right to decide tie-breaks)
30. botright: Colour in the bottom-left corner (moving left to decide tie-breaks)
```

Exercise - read the help for `read_csv` and figure out how to load the data with the above variable names. One pitfall to note - with `header=None` pandas generated variable names starting from 0, but the above list starts from 1...

## Your assignment - pick a dataset and do something like the above

This is purposely open-ended - you can pick any data set you wish. It is highly advised you pick a dataset from UCI or a similar "clean" source.

If you get that done and want to try more challenging or exotic things, go for it! Use documentation as illustrated above, and follow the 20-minute rule (that is - ask for help if you're stuck).

If you have loaded a few traditional datasets, see the following section for suggested stretch goals.

For this assignment, I am pulling a dataset from the UCI repository. 

**Dataset:** [Sao Paolo Traffic](https://archive.ics.uci.edu/ml/datasets/Behavior+of+the+urban+traffic+of+the+city+of+Sao+Paulo+in+Brazil)

It is a relatively new dataset so they have no information on NA values or cleanliness.

In [0]:
import numpy as np
import pandas as pd 
from google.colab import files

In [2]:
uploaded = files.upload()

Saving Behavior of the urban traffic of the city of Sao Paulo in Brazil.csv to Behavior of the urban traffic of the city of Sao Paulo in Brazil.csv


In [56]:
df = pd.read_csv('Behavior of the urban traffic of the city of Sao Paulo in Brazil.csv', sep = None) #Having some seperator issues. 
df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support sep=None with delim_whitespace=False; you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


Hour (Coded)  Immobilized bus  Broken Truck  Vehicle excess  \
0             1                0             0               0   
1             2                0             0               0   
2             3                0             0               0   
3             4                0             0               0   
4             5                0             0               0   

   Accident victim  Running over  Fire vehicles  Occurrence involving freight  \
0                0             0              0                             0   
1                0             0              0                             0   
2                0             0              0                             0   
3                0             0              0                             0   
4                0             0              0                             0   

   Incident involving dangerous freight  Lack of electricity  Fire  \
0                                     0                    0     0   
1                                     0                    0     0   
2                                     0                    0     0   
3                                     0                    0     0   
4                                     0                    0     0   

   Point of flooding  Manifestations  Defect in the network of trolleybuses  \
0                  0               0                                      0   
1                  0               0                                      0   
2                  0               0                                      0   
3                  0               0                                      0   
4                  0               0                                      0   

   Tree on the road  Semaphore off  Intermittent Semaphore  \
0                 0              0                       0   
1                 0              0                       0   
2                 0              0                       0   
3                 0              0                       0   
4                 0              0                       0   

  Slowness in traffic (%)  
0                     4,1  
1                     6,6  
2                     8,7  
3                     9,2  
4                    11,1

In [17]:
df.shape

(135, 18)

Well that is a pretty clean dataset. I'm going pull in another one to clean up. 



In [21]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/hepatitis/hepatitis.data')
print(df.shape)
df.head()

(154, 20)


2  30  2.1  1  2.2 2.3 2.4 2.5 1.1 2.6 2.7 2.8 2.9 2.10  1.00   85   18  \
0  2  50    1  1    2   1   2   2   1   2   2   2   2    2  0.90  135   42   
1  2  78    1  2    2   1   2   2   2   2   2   2   2    2  0.70   96   32   
2  2  31    1  ?    1   2   2   2   2   2   2   2   2    2  0.70   46   52   
3  2  34    1  2    2   2   2   2   2   2   2   2   2    2  1.00    ?  200   
4  2  34    1  2    2   2   2   2   2   2   2   2   2    2  0.90   95   28   

   4.0   ?  1.2  
0  3.5   ?    1  
1  4.0   ?    1  
2  4.0  80    1  
3  4.0   ?    1  
4  4.0  75    1

**Looks like the dataset has "?" indicated NAN entries and we need to add some headers. Here is the index of header from the website:**

 *Attribute Information:*

1. Class: DIE, LIVE 
2. AGE: 10, 20, 30, 40, 50, 60, 70, 80 
3. SEX: male, female 
4. STEROID: no, yes 
5. ANTIVIRALS: no, yes 
6. FATIGUE: no, yes 
7. MALAISE: no, yes 
8. ANOREXIA: no, yes 
9. LIVER BIG: no, yes 
10. LIVER FIRM: no, yes 
11. SPLEEN PALPABLE: no, yes 
12. SPIDERS: no, yes 
13. ASCITES: no, yes 
14. VARICES: no, yes 
15. BILIRUBIN: 0.39, 0.80, 1.20, 2.00, 3.00, 4.00 
-- see the note below 
16. ALK PHOSPHATE: 33, 80, 120, 160, 200, 250 
17. SGOT: 13, 100, 200, 300, 400, 500, 
18. ALBUMIN: 2.1, 3.0, 3.8, 4.5, 5.0, 6.0 
19. PROTIME: 10, 20, 30, 40, 50, 60, 70, 80, 90 
20. HISTOLOGY: no, yes 

The BILIRUBIN attribute appears to be continuously-valued. I checked this with the donater, Bojan Cestnik, who replied: 

About the hepatitis database and BILIRUBIN problem I would like to say the following: BILIRUBIN is continuous attribute (= the number of it's "values" in the ASDOHEPA.DAT file is negative!!!); "values" are quoted because when speaking about the continuous attribute there is no such thing as all possible values. However, they represent so called "boundary" values; according to these "boundary" values the attribute can be discretized. At the same time, because of the continious attribute, one can perform some other test since the continuous information is preserved. I hope that these lines have at least roughly answered your question text*

**Now to add headers and replace "?" with NAN so we can deal with those normally.**



In [113]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/hepatitis/hepatitis.data',na_values = ['?'], header = None, names=['Class','Age','Sex','Steroid','Antivirals',
                                                                                                                      'Fatigue','Malaise','Anorexia','Liver_Big','Liver_Firm',
                                                                                                                      'Spleen_Palpable','Spiders','Ascites','Varices','Bilrubin',
                                                                                                                      'Alk_Phosphate','SGOT','Albumin','Protime','Histology'])
df.head()

Class  Age  Sex  Steroid  Antivirals  Fatigue  Malaise  Anorexia  \
0      2   30    2      1.0           2      2.0      2.0       2.0   
1      2   50    1      1.0           2      1.0      2.0       2.0   
2      2   78    1      2.0           2      1.0      2.0       2.0   
3      2   31    1      NaN           1      2.0      2.0       2.0   
4      2   34    1      2.0           2      2.0      2.0       2.0   

   Liver_Big  Liver_Firm  Spleen_Palpable  Spiders  Ascites  Varices  \
0        1.0         2.0              2.0      2.0      2.0      2.0   
1        1.0         2.0              2.0      2.0      2.0      2.0   
2        2.0         2.0              2.0      2.0      2.0      2.0   
3        2.0         2.0              2.0      2.0      2.0      2.0   
4        2.0         2.0              2.0      2.0      2.0      2.0   

   Bilrubin  Alk_Phosphate   SGOT  Albumin  Protime  Histology  
0       1.0           85.0   18.0      4.0      NaN          1  
1       0.9          135.0   42.0      3.5      NaN          1  
2       0.7           96.0   32.0      4.0      NaN          1  
3       0.7           46.0   52.0      4.0     80.0          1  
4       1.0            NaN  200.0      4.0      NaN          1

In [0]:
from statistics import mode


In [0]:
def fill_na():
  
  df.Albumin = df.Albumin.fillna(np.mean(df.Albumin))
  df.Alk_Phosphate = df.Alk_Phosphate.fillna(mode(df.Alk_Phosphate))
  df.SGOT = df.SGOT.fillna(mode(df.Alk_Phosphate))
  df.Bilrubin = df.Bilrubin.fillna(np.mean(df.Bilrubin))
  df.Protime = df.Protime.fillna(float(60))
  
  
  
  
fill_na()

In [118]:
df.isnull().sum()

Class               0
Age                 0
Sex                 0
Steroid             1
Antivirals          0
Fatigue             1
Malaise             1
Anorexia            1
Liver_Big          10
Liver_Firm         11
Spleen_Palpable     5
Spiders             5
Ascites             5
Varices             5
Bilrubin            0
Alk_Phosphate       0
SGOT                0
Albumin             0
Protime             0
Histology           0
dtype: int64

In [51]:
df.dtypes

Class                int64
Age                  int64
Sex                  int64
Steroid            float64
Antivirals           int64
Fatigue            float64
Malaise            float64
Anorexia           float64
Liver_Big          float64
Liver_Firm         float64
Spleen_Palpable    float64
Spiders            float64
Ascites            float64
Varices            float64
Bilrubin           float64
Alk_Phosphate      float64
SGOT               float64
Albumin            float64
Protime            float64
Histology            int64
dtype: object

In [54]:
df.Spleen_Palpable

0      2.0
1      2.0
2      2.0
3      2.0
4      2.0
5      2.0
6      1.0
7      2.0
8      2.0
9      2.0
10     2.0
11     2.0
12     2.0
13     2.0
14     2.0
15     2.0
16     2.0
17     2.0
18     2.0
19     2.0
20     2.0
21     1.0
22     2.0
23     2.0
24     2.0
25     2.0
26     2.0
27     2.0
28     2.0
29     2.0
      ... 
125    2.0
126    2.0
127    1.0
128    2.0
129    2.0
130    2.0
131    1.0
132    2.0
133    2.0
134    2.0
135    1.0
136    2.0
137    2.0
138    2.0
139    2.0
140    2.0
141    1.0
142    1.0
143    1.0
144    2.0
145    2.0
146    1.0
147    NaN
148    2.0
149    2.0
150    2.0
151    2.0
152    2.0
153    1.0
154    1.0
Name: Spleen_Palpable, Length: 155, dtype: float64

## Stretch Goals - Other types and sources of data

Not all data comes in a nice single file - for example, image classification involves handling lots of image files. You still will probably want labels for them, so you may have tabular data in addition to the image blobs - and the images may be reduced in resolution and even fit in a regular csv as a bunch of numbers.

If you're interested in natural language processing and analyzing text, that is another example where, while it can be put in a csv, you may end up loading much larger raw data and generating features that can then be thought of in a more standard tabular fashion.

Overall you will in the course of learning data science deal with loading data in a variety of ways. Another common way to get data is from a database - most modern applications are backed by one or more databases, which you can query to get data to analyze. We'll cover this more in our data engineering unit.

How does data get in the database? Most applications generate logs - text files with lots and lots of records of each use of the application. Databases are often populated based on these files, but in some situations you may directly analyze log files. The usual way to do this is with command line (Unix) tools - command lines are intimidating, so don't expect to learn them all at once, but depending on your interests it can be useful to practice.

One last major source of data is APIs: https://github.com/toddmotto/public-apis

API stands for Application Programming Interface, and while originally meant e.g. the way an application interfaced with the GUI or other aspects of an operating system, now it largely refers to online services that let you query and retrieve data. You can essentially think of most of them as "somebody else's database" - you have (usually limited) access.

*Stretch goal* - research one of the above extended forms of data/data loading. See if you can get a basic example working in a notebook. Image, text, or (public) APIs are probably more tractable - databases are interesting, but there aren't many publicly accessible and they require a great deal of setup.